### PR TITLE
Migrate to puppet4 datatypes

### PR DESCRIPTION
**Pull request by bastelfreak**
The goal of this is to get rid of most of the validate_* calls, not to introduce datatypes to _all_ params.

### CODE REVIEW

File: manifests/server/config.pp [hunk 1]
```diff
@@ -49,40 +49,40 @@
         user        => $user,
         auth_method => 'ident',
         auth_option => $local_auth_option,
-        order       => '001',
+        order       => 001,
```
Comment (binford2k) [inline manifests/server/config.pp]: This is probably a bad pattern to get into. The leading 0 makes it octal, which is fine by this, but not for other numbers.

```
$ puppet apply -e 'notice(002)'
Notice: Scope(Class[main]): 2
Notice: Compiled catalog for ganymede.local in environment production in 0.15 seconds
Notice: Applied catalog in 0.03 seconds
$ puppet apply -e 'notice(009)'
Error: Could not parse for environment production: Not a valid octal number 009  at line 1:8 on node ganymede.local
```

File: manifests/client.pp [hunk 2]
```diff
@@ -1,12 +1,10 @@
 # Install client cli tool. See README.md for more details.
 class postgresql::client (
-  $file_ensure    = 'file',
-  $validcon_script_path  = $postgresql::params::validcon_script_path,
-  $package_name   = $postgresql::params::client_package_name,
-  $package_ensure = 'present'
+  Enum['file', 'absent'] $file_ensure                 = 'file',
```
Comment (alexjfisher) [inline manifests/client.pp]: Bad style, but some users might be using 'present'?

File: manifests/lib/devel.pp [hunk 3]
```diff
@@ -1,13 +1,11 @@
 # This class installs postgresql development libraries. See README.md for more
 # details.
 class postgresql::lib::devel(
-  $package_name   = $postgresql::params::devel_package_name,
-  $package_ensure = 'present',
-  $link_pg_config = $postgresql::params::link_pg_config
+  String $package_name                               = $postgresql::params::devel_package_name,
+  Enum['present', 'absent', 'latest']$package_ensure = 'present',
```
Comment (alexjfisher) [inline manifests/lib/devel.pp]: Missing space?

Comment (alexjfisher) [inline manifests/lib/devel.pp]: Also need 'installed'?

File: manifests/lib/docs.pp [hunk 4]
```diff
@@ -1,12 +1,10 @@
 # This class installs the postgresql-docs See README.md for more
 # details.
 class postgresql::lib::docs (
-  $package_name   = $postgresql::params::docs_package_name,
-  $package_ensure = 'present',
+  String $package_name                                = $postgresql::params::docs_package_name,
+  Enum['present', 'absent', 'latest'] $package_ensure = 'present',
```
Comment (alexjfisher) [inline manifests/lib/docs.pp]: And 'installed'?

File: manifests/lib/java.pp [hunk 5]
```diff
@@ -1,12 +1,10 @@
 # This class installs the postgresql jdbc connector. See README.md for more
 # details.
 class postgresql::lib::java (
-  $package_name   = $postgresql::params::java_package_name,
-  $package_ensure = 'present'
+  String $package_name                                = $postgresql::params::java_package_name,
+  Enum['present', 'absent', 'latest'] $package_ensure = 'present'
```
Comment (alexjfisher) [inline manifests/lib/java.pp]: Installed?

File: manifests/server/contrib.pp [hunk 6]
```diff
@@ -1,9 +1,8 @@
 # Install the contrib postgresql packaging. See README.md for more details.
 class postgresql::server::contrib (
-  $package_name   = $postgresql::params::contrib_package_name,
-  $package_ensure = 'present'
+  String $package_name                                = $postgresql::params::contrib_package_name,
+  Enum['present', 'absent', 'latest'] $package_ensure = 'present'
```
Comment (alexjfisher) [inline manifests/server/contrib.pp]: Installed?

File: manifests/server/grant.pp [hunk 7]
```diff
@@ -1,16 +1,17 @@
...
-  $port             = $postgresql::server::port,
-  $onlyif_exists    = false,
-  $connect_settings = $postgresql::server::default_connect_settings,
+  String $role,
+  String $db,
+  Optional[String] $privilege = undef,
+  String $object_type         = 'database',
+  $object_name                = undef,
+  String $psql_db             = $postgresql::server::default_database,
+  String$psql_user            = $postgresql::server::user,
```
Comment (alexjfisher) [inline manifests/server/grant.pp]: Missing space.

File: manifests/server/postgis.pp [hunk 8]
```diff
@@ -1,9 +1,8 @@
 # Install the postgis postgresql packaging. See README.md for more details.
 class postgresql::server::postgis (
-  $package_name   = $postgresql::params::postgis_package_name,
-  $package_ensure = 'present'
+  String $package_name                                = $postgresql::params::postgis_package_name,
+  Enum['present', 'absent', 'latest'] $package_ensure = 'present'
```
Comment (alexjfisher) [inline manifests/server/postgis.pp]: Installed?

File: metadata.json [hunk 9]
```diff
@@ -8,9 +8,9 @@
   "project_page": "https://github.com/puppetlabs/puppetlabs-postgresql",
   "issues_url": "https://tickets.puppetlabs.com/browse/MODULES",
   "dependencies": [
-    {"name":"puppetlabs/stdlib","version_requirement":"4.x"},
-    {"name":"puppetlabs/apt","version_requirement":">=2.0.0 <3.0.0"},
-    {"name":"puppetlabs/concat","version_requirement":">= 1.1.0 <3.0.0"}
+    {"name":"puppetlabs/stdlib","version_requirement":">= 4.13.1 < 5.0.0"},
+    {"name":"puppetlabs/apt","version_requirement":">= 2.0.0 < 3.0.0"},
+    {"name":"puppetlabs/concat","version_requirement":">= 1.1.0 < 3.0.0"}
```
Comment (alexjfisher) [inline metadata.json]: Is this still correct? (New versions of concat have been released since this PR was opened)

File: spec/unit/defines/server/pg_hba_rule_spec.rb [hunk 10]
```diff
@@ -190,7 +143,7 @@ class { 'postgresql::server': }
       it do
         is_expected.to contain_concat__fragment('pg_hba_rule_test').with({
           :content => /local\s+all\s+all\s+0\.0\.0\.0\/0\s+peer/
-	})
+      })
```
Comment (alexjfisher) [inline spec/unit/defines/server/pg_hba_rule_spec.rb]: Alignment doesn't look right here

File: manifests/client.pp [hunk 11]
```diff
@@ -1,12 +1,10 @@
...
+  Stdlib::Absolutepath $validcon_script_path          = $postgresql::params::validcon_script_path,
+  String $package_name                                = $postgresql::params::client_package_name,
```
Comment (hasegeli) [inline manifests/client.pp]: It cannot be an empty string.  `String[1]` would be more appropriate.

File: manifests/client.pp [hunk 12]
```diff
@@ -1,12 +1,10 @@
...
+  Enum['present', 'absent', 'latest'] $package_ensure = 'present'
```
Comment (hasegeli) [inline manifests/client.pp]: We can define it as `Postgresql::Package_ensure` not to repeat ourself.

File: manifests/server/config.pp [hunk 13]
```diff
@@ -49,40 +49,40 @@
         user        => $user,
         auth_method => 'ident',
         auth_option => $local_auth_option,
-        order       => '001',
+        order       => 1,
```
Comment (hasegeli) [inline manifests/server/config.pp]: This is actually working as a string on the Concat module.  `2` would be sorted after after `10`.

File: manifests/server/grant.pp [hunk 14]
```diff
@@ -1,16 +1,17 @@
...
-  $psql_db          = $postgresql::server::default_database,
-  $psql_user        = $postgresql::server::user,
-  $port             = $postgresql::server::port,
-  $onlyif_exists    = false,
-  $connect_settings = $postgresql::server::default_connect_settings,
+  String $role,
+  String $db,
+  Optional[String] $privilege = undef,
+  String $object_type         = 'database',
+  $object_name                = undef,
```
Comment (hasegeli) [inline manifests/server/grant.pp]: Can't we define this one as `String[1]`?

File: manifests/server/grant.pp [hunk 15]
```diff
@@ -1,16 +1,17 @@
...
+  String $psql_db             = $postgresql::server::default_database,
+  String$psql_user            = $postgresql::server::user,
+  Integer $port               = $postgresql::server::port,
+  Boolean $onlyif_exists      = false,
+  Hash $connect_settings      = $postgresql::server::default_connect_settings,
```
Comment (hasegeli) [inline manifests/server/grant.pp]: Can't we specify what is in the `Hash`?

File: manifests/server/grant.pp [hunk 16]
```diff
@@ -1,16 +1,17 @@
...
-  $onlyif_exists    = false,
-  $connect_settings = $postgresql::server::default_connect_settings,
+  String $role,
+  String $db,
+  Optional[String] $privilege = undef,
+  String $object_type         = 'database',
+  $object_name                = undef,
+  String $psql_db             = $postgresql::server::default_database,
+  String$psql_user            = $postgresql::server::user,
+  Integer $port               = $postgresql::server::port,
```
Comment (hasegeli) [inline manifests/server/grant.pp]: It should be greater than 0.  I am not sure if there is a portable max limit.

File: manifests/client.pp (hunk 2 above)
Comment (bastelfreak) [inline manifests/client.pp]: the next release will be a major one, so should be prohibit the bad practice?

File: manifests/lib/devel.pp (hunk 3 above)
Comment (bastelfreak) [inline manifests/lib/devel.pp]: fixed

File: manifests/lib/java.pp (hunk 5 above)
Comment (bastelfreak) [inline manifests/lib/java.pp]: fixed

File: manifests/server/contrib.pp (hunk 6 above)
Comment (bastelfreak) [inline manifests/server/contrib.pp]: fixed

File: manifests/server/grant.pp (hunk 7 above)
Comment (bastelfreak) [inline manifests/server/grant.pp]: fixed

File: manifests/server/postgis.pp (hunk 8 above)
Comment (bastelfreak) [inline manifests/server/postgis.pp]: fixed

File: metadata.json (hunk 9 above)
Comment (bastelfreak) [inline metadata.json]: I'm happily using it with concat4. but bumping this shouldn't be part of this PR.

File: spec/unit/defines/server/pg_hba_rule_spec.rb (hunk 10 above)
Comment (bastelfreak) [inline spec/unit/defines/server/pg_hba_rule_spec.rb]: fixed

File: manifests/server/grant.pp (hunk 14 above)
Comment (bastelfreak) [inline manifests/server/grant.pp]: Which atribute do you mean, object_name? This one is optional.

File: manifests/client.pp (hunk 11 above)
Comment (bastelfreak) [inline manifests/client.pp]: fixed

File: manifests/server/grant.pp (hunk 14 above)
Comment (hasegeli) [inline manifests/server/grant.pp]: `Optional[String[1]]` maybe?

Comment (bastelfreak) [inline manifests/server/grant.pp]: sounds good, will do.

File: manifests/server/grant_role.pp [hunk 17]
```diff
@@ -1,21 +1,13 @@
 # Define for granting membership to a role. See README.md for more information
 define postgresql::server::grant_role (
-  $group,
-  $role             = $name,
-  $ensure           = 'present',
+  String $group,
```
Comment (ekohl) [inline manifests/server/grant_role.pp]: I believe this accepts empty strings. Maybe ```String[1]```? Same for role below.

File: manifests/server/pg_hba_rule.pp [hunk 18]
```diff
@@ -1,19 +1,19 @@
...
-  $user,
-  $auth_method,
-  $address     = undef,
-  $description = 'none',
-  $auth_option = undef,
-  $order       = '150',
+  Enum['local', 'host', 'hostssl', 'hostnossl'] $type,
+  String $database,
+  String $user,
+  String $auth_method,
```
Comment (ekohl) [inline manifests/server/pg_hba_rule.pp]: In the validation code there's some regex validation with ```validate_re```. I believe you can simplify that to ```assert_type(Enum[$allowed_auth_methods], $auth_method)``` but I'm not 100% sure.

File: manifests/client.pp (hunk 2 above)
Comment (hunner) [inline manifests/client.pp]: 👍

File: manifests/lib/devel.pp [hunk 19]
```diff
@@ -1,13 +1,11 @@
 # This class installs postgresql development libraries. See README.md for more
 # details.
 class postgresql::lib::devel(
-  $package_name   = $postgresql::params::devel_package_name,
-  $package_ensure = 'present',
-  $link_pg_config = $postgresql::params::link_pg_config
+  String $package_name                                             = $postgresql::params::devel_package_name,
+  Enum['present', 'absent', 'latest', 'installed'] $package_ensure = 'present',
```
Comment (hunner) [inline manifests/lib/devel.pp]: Here as well.

File: manifests/lib/docs.pp [hunk 20]
```diff
@@ -1,12 +1,10 @@
 # This class installs the postgresql-docs See README.md for more
 # details.
 class postgresql::lib::docs (
-  $package_name   = $postgresql::params::docs_package_name,
-  $package_ensure = 'present',
+  String $package_name                                             = $postgresql::params::docs_package_name,
+  Enum['present', 'absent', 'latest', 'installed'] $package_ensure = 'present',
```
Comment (hunner) [inline manifests/lib/docs.pp]: Here as well.

File: manifests/lib/java.pp [hunk 21]
```diff
@@ -1,12 +1,10 @@
 # This class installs the postgresql jdbc connector. See README.md for more
 # details.
 class postgresql::lib::java (
-  $package_name   = $postgresql::params::java_package_name,
-  $package_ensure = 'present'
+  String $package_name                                             = $postgresql::params::java_package_name,
+  Enum['present', 'absent', 'latest', 'installed'] $package_ensure = 'present'
```
Comment (hunner) [inline manifests/lib/java.pp]: Here as well.

File: manifests/server/config.pp (hunk 13 above)
Comment (hunner) [inline manifests/server/config.pp]: Can we change the concat resource to `order => 'numeric'` to help? I think that makes more sense anyway.

File: manifests/server/contrib.pp [hunk 22]
```diff
@@ -1,9 +1,8 @@
 # Install the contrib postgresql packaging. See README.md for more details.
 class postgresql::server::contrib (
-  $package_name   = $postgresql::params::contrib_package_name,
-  $package_ensure = 'present'
+  String $package_name                                = $postgresql::params::contrib_package_name,
+  Enum['present', 'absent', 'latest', 'installed'] $package_ensure = 'present'
```
Comment (hunner) [inline manifests/server/contrib.pp]: Here as well.

File: manifests/server/pg_hba_rule.pp (hunk 18 above)
Comment (hunner) [inline manifests/server/pg_hba_rule.pp]: I think that's valid and useful.

File: manifests/server/postgis.pp [hunk 23]
```diff
@@ -1,9 +1,8 @@
 # Install the postgis postgresql packaging. See README.md for more details.
 class postgresql::server::postgis (
-  $package_name   = $postgresql::params::postgis_package_name,
-  $package_ensure = 'present'
+  String $package_name                                             = $postgresql::params::postgis_package_name,
+  Enum['present', 'absent', 'latest', 'installed'] $package_ensure = 'present'
```
Comment (hunner) [inline manifests/server/postgis.pp]: Here as well.

File: manifests/server/config.pp (hunk 13 above)
Comment (bastelfreak) [inline manifests/server/config.pp]: done

File: manifests/server/grant_role.pp (hunk 17 above)
Comment (bastelfreak) [inline manifests/server/grant_role.pp]: fixed

File: manifests/server/pg_hba_rule.pp (hunk 18 above)
Comment (bastelfreak) [inline manifests/server/pg_hba_rule.pp]: fixed
